### PR TITLE
test/devux: helper function for assertions conditional on panics

### PR DESCRIPTION
**Pull request by stackman27**
Closes: #1946

## What is the purpose of the change

Adds a helper function for conditioning on Panics 
Realized that there is already a helper function in `x/gamm/pool-models/balancer/util_test.go` so i just moved that under `osmoutils`

## Brief Changelog
n/a

## Testing and Verifying
n/a

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Comment by stackman27** [comment]
We could use something similar for `conditionalErrors`, but it is going to be a little more complex. These are 3 ways that i've found `expectError` being used ;

1. Basic implementation 
```
if tc.expectErr {
    require.Error(t, err)
} else {
    require.NoError(t, err)
}
```
2. Error with message 
```
if test.expectPass {
   require.NoError(t, test.msg.ValidateBasic(), "test: %v", test.name)
} else {
   require.Error(t, test.msg.ValidateBasic(), "test: %v", test.name)
}
```

3. Checks after NoError 
```
if tc.expectErr {
	require.Error(t, err)
} else {
	require.NoError(t, err)
	require.NotNil(t, resp)
	require.Equal(t, tc.expectAdmin, resp.Admin)
}
```

4. Check that halts after error 
```
if tc.expErr {
    require.Error(t, gotErr)
    return
}
require.NoError(t, gotErr)
```

If this is something that we should add. Please let me know and i can add it in this PR :)

File: osmoutils/cli_helpers.go [hunk 1]
```diff
@@ -47,3 +49,16 @@ func ParseSdkIntFromString(s string, separator string) ([]sdk.Int, error) {
...
+// returns true if sut panics and false it it doesnot
+func ConditionalPanic(t *testing.T, expectPanic bool, sut func()) bool {
+	if expectPanic {
+		require.Panics(t, sut)
+		return true
+	}
+
+	require.NotPanics(t, sut)
+	return false
+}
```
Comment (ValarDragon) [inline osmoutils/cli_helpers.go:64]:
```suggestion
func ConditionalPanic(t *testing.T, expectPanic bool, sut func()) {
	if expectPanic {
		require.Panics(t, sut)
	}

	require.NotPanics(t, sut)
}
```

File: x/mint/keeper/genesis_test.go [hunk 2]
```diff
@@ -111,17 +112,11 @@ func (suite *KeeperTestSuite) TestMintInitGenesis() {
 			originalVestingCoins := bankKeeper.GetBalance(ctx, developerAccount, tc.mintDenom)
 
 			// Test.
-			if tc.expectPanic {
-				suite.Panics(func() {
-					mintKeeper.InitGenesis(ctx, tc.mintGenesis)
-				})
+
+			if osmoutils.ConditionalPanic(suite.T(), tc.expectPanic || false, func() { mintKeeper.InitGenesis(ctx, tc.mintGenesis) }) {
```
Comment (ValarDragon) [inline x/mint/keeper/genesis_test.go:116]: What is tc.expectPanic || false for?

Comment (ValarDragon) [reply x/mint/keeper/genesis_test.go:116]: Also I recommend we change this to \
```suggestion
			osmoutils.ConditionalPanic(suite.T(), tc.expectPanic, func() { mintKeeper.InitGenesis(ctx, tc.mintGenesis) })
			if tc.expectPanic {
			    return
			}
```

**Comment by ValarDragon** [comment]
Thanks for adding this! Would recommend we don't do this for expect pass atm

File: osmoutils/cli_helpers.go (hunk 1 above)
Comment (stackman27) [reply osmoutils/cli_helpers.go:64]: @ValarDragon Would this be; cause i'm getting an error without the return
```
func ConditionalPanic(t *testing.T, expectPanic bool, sut func()) {
	if expectPanic {
		require.Panics(t, sut)
                 return 
	}
	require.NotPanics(t, sut)
}
```

File: x/mint/keeper/genesis_test.go (hunk 2 above)
Comment (stackman27) [reply x/mint/keeper/genesis_test.go:116]: Oops i forgot that an unassigned bool is basically false

**Comment by stackman27** [comment]
> Thanks for adding this! Would recommend we don't do this for expect pass atm

Thank you @ValarDragon! Resolved your comments

**Review by mattverse — CHANGES_REQUESTED** [review]
Logic and feature LGTM!

I'd prefer creating a new file `test_helpers.go` in osmo utils instead of having the method within `cli_helpers`

File: osmoutils/cli_helpers.go [hunk 3]
```diff
@@ -47,3 +49,15 @@ func ParseSdkIntFromString(s string, separator string) ([]sdk.Int, error) {
 	}
 	return parsedInts, nil
 }
+
+// ConditionalPanic checks if expectPanic is true, asserts that sut (system under test)
+// panics. If expectPanic is false, asserts that sut does not panic.
+// returns true if sut panics and false it it doesnot
```
Comment (p0mvn) [inline osmoutils/cli_helpers.go:55]:
```suggestion
// returns true if sut panics and false if it does not
```

**Review by p0mvn — APPROVED** [review]
Sweet, LGTM once Matt's suggestion to move to a different file is addressed

File: osmoutils/cli_helpers.go [hunk 4]
```diff
@@ -47,3 +49,15 @@ func ParseSdkIntFromString(s string, separator string) ([]sdk.Int, error) {
 	}
 	return parsedInts, nil
 }
+
+// ConditionalPanic checks if expectPanic is true, asserts that sut (system under test)
+// panics. If expectPanic is false, asserts that sut does not panic.
+// returns true if sut panics and false if it does not
+func ConditionalPanic(t *testing.T, expectPanic bool, sut func()) {
```
Comment (alexanderbez) [inline osmoutils/cli_helpers.go:56]: I'm not sure if this is the correct file for such a helper to exist. Do have a place for more general shared testing auxiliary stuff?

Comment (stackman27) [reply osmoutils/cli_helpers.go]: created a new file called `test_helpers.go` as matt suggested

**Comment by stackman27** [comment]
Good to go!

**Review by ValarDragon — APPROVED** [review]

**Pull request merged**